[Dependency Analysis] Remove Unused Dependencies

ParaskP7 opened the pull request:
Project Thread: paaHJt-6YV-p2

### Description

Based on the `unused` related advises generated by the [Dependency Analysis](https://github.com/autonomousapps/dependency-analysis-gradle-plugin) Gradle plugin, this PR removes all unused dependencies from this project.

FYI: As part of this change, the [Dependency Analysis](https://github.com/autonomousapps/dependency-analysis-gradle-plugin) Gradle plugin got also updated to its latest [1.33.0](https://github.com/autonomousapps/dependency-analysis-gradle-plugin/blob/main/CHANGELOG.md#version-1330) version.

### Testing Steps

1. Tooling:
    - Run the `./gradlew buildHealth` task and verify that there is no `unused` related advise remaining within both reports, JSON and txt included:
        - `build-health-report.json` -> Search for `unusedCount`
        - `build-health-report.txt` -> Search for `unused`
    - Check the manually triggered [scheduled build](https://buildkite.com/automattic/automattic-tracks-android/builds/317) and verify that everything works as expected with this newer version of the [Dependency Analysis](https://github.com/autonomousapps/dependency-analysis-gradle-plugin) Gradle plugin ([1.33.0](https://github.com/autonomousapps/dependency-analysis-gradle-plugin/blob/main/CHANGELOG.md#version-1330)).
3. Testing:
    - Smoke test the sample tracks app, try out all available screens and functionality. Verify everything is working as expected.
    - Also, if you want to be thorough about reviewing the changes, you could quickly smoke test, either the [JP/WPAndroid](https://github.com/wordpress-mobile/WordPress-Android) and/or [WCAndroid](https://github.com/woocommerce/woocommerce-android), with this version of `Tracks`, or via composite builds, and see if it works as expected.

ParaskP7 committed:
Deps: Update dependency analysis to 1.33.0

Release Notes: https://github.com/autonomousapps/
dependency-analysis-gradle-plugin/blob/main/CHANGELOG.md#version-1330

ParaskP7 committed:
Deps: Remove squareup okhttp3 dependency from automattic tracks

This dependency was added as part of this
(9b00ec5a41dc6cb260ad8c937f9953131bd3bb62) commit, which added the
'CrashLoggingOkHttpInterceptorProvider' class on the project. But then,
as part of this (24ae7506517ea59a8c725fbaba222c07720fbcfc) other commit,
this 'CrashLoggingOkHttpInterceptorProvider' class got moved into the
newly created 'crashlogging' module. As such, this dependency is no
longer needed on 'AutomatticTracks'.

------------------------------------------------------------------------

This removal was suggested by the dependency analysis report, see below:

Advice for :AutomatticTracks
Unused dependencies which should be removed:
  implementation 'com.squareup.okhttp3:okhttp:4.9.0'

ParaskP7 committed:
Build: Update dependency analysis configuration on benchmark

This 'benchmark' module is a special module that was automatically
generated from the Android Studio wizard
(828fb43f15f63f3f25fcf699568cf0f5e2e26205) for basic track method
benchmarking purposes only (7a302f1151351d57553648e8b397da11af9fdeba).
Ignoring this Kotlin Stdlib advise instead of trying to act on this
advise seems reasonable, especially since there is no Kotlin code on the
main source-set whatsoever.

As such this advice is instead suppressed and the 'onUnusedDependencies'
exclusion configuration were used to achieve that.

------------------------------------------------------------------------

Advice for :benchmark
Unused dependencies which should be removed:
  api 'org.jetbrains.kotlin:kotlin-stdlib:1.9.24'

ParaskP7 committed:
Deps: Replace sentry android okhttp with sentry okhttp in crashlogging

Removing the 'sentry-android-okhttp' dependency wasn't enough by itself
as the build was then failing. Adding the transient 'sentry-okhttp'
dependency was actually required in order to make the build successful
again.

------------------------------------------------------------------------

This replacement was suggested by the dependency analysis report,
see below:

Advice for :crashlogging
Unused dependencies which should be removed:
  implementation 'io.sentry:sentry-android-okhttp:7.14.0'
  ...

These transitive dependencies should be declared directly:
  ...
  implementation 'io.sentry:sentry-okhttp:7.14.0'
  ...

ParaskP7 committed:
Deps: Replace sentry android with sentry android core on crashlogging

Removing the 'sentry-android' dependency wasn't enough by itself as the
build was then failing. Adding the transient 'sentry-android-core'
dependency was actually required in order to make the build successful
again.

------------------------------------------------------------------------

This replacement was suggested by the dependency analysis report,
see below:

Advice for :crashlogging
Unused dependencies which should be removed:
  implementation 'io.sentry:sentry-android:7.14.0'

These transitive dependencies should be declared directly:
  ...
  implementation 'io.sentry:sentry-android-core:7.14.0'
  ...

ParaskP7 committed:
Deps: Remove androidx core dependency from sample tracks app

This dependency is completely unused.

------------------------------------------------------------------------

This removal was suggested by the dependency analysis report, see below:

Advice for :sampletracksapp
Unused dependencies which should be removed:
  implementation 'androidx.core:core-ktx:1.3.2'
  ...

ParaskP7 committed:
Build: Update dependency analysis configuration on sample tracks app

This 'unused' related advise for the 'sampletracksapp' module is
actually correct and 'AutomatticTracks' can be removed as a dependency
since it is actually unused and not being utilized from within the
sample tracks app.

However, I chose to keep it there and suppress the advise instead in
order to preserve the sample app default configuration, its purpose
being to test all those individual modules:
- AutomatticTracks
- crashlogging

As such this advice is instead suppressed and the 'onUnusedDependencies'
exclusion configuration were used to achieve that.

------------------------------------------------------------------------

Advice for :sampletracksapp
Unused dependencies which should be removed:
  implementation project(':AutomatticTracks')

ParaskP7 committed:
Analysis: Suppress and ignore missing activity for tracking benchmark

Error Message: "Class referenced in the manifest,
com.automattic.tracks.benchmark.TrackingMethodsBenchmark$SampleActivity,
was not found in the project or the libraries"

ParaskP7 committed:
Build: Add missing lint configuration to benchmark module

ParaskP7 committed:
Build: Add missing lint configuration to crashlogging module

ParaskP7 committed:
Analysis: Disable obsolete lint custom check on crashlogging

Error Message: "Error: Library lint checks reference invalid APIs;
these checks will be skipped!

Lint found an issue registry (androidx.compose.runtime.lint.
RuntimeIssueRegistry)
which contains some references to invalid API:
org.jetbrains.kotlin.analysis.api.session.KtAnalysisSessionProvider:
org.jetbrains.kotlin.analysis.api.lifetime.KtLifetimeTokenFactory
getTokenFactory()
(Referenced from androidx/compose/runtime/lint/
AutoboxingStateCreationDetector.class)

Therefore, this lint check library is not included
in analysis. This affects the following lint checks:
AutoboxingStateValueProperty
AutoboxingStateCreation
CoroutineCreationDuringComposition
FlowOperatorInvokedInComposition
ComposableLambdaParameterNaming
ComposableLambdaParameterPosition
ComposableNaming
StateFlowValueCalledInComposition
CompositionLocalNaming
MutableCollectionMutableState
ProduceStateDoesNotAssignValue
RememberReturnType
OpaqueUnitKey
UnrememberedMutableState"

ParaskP7 committed:
CI: Run lint on ci for both the debug and release variants

wzieba commented on benchmark/build.gradle.kts:
```diff
@@ -57,3 +62,12 @@ dependencies {
     androidTestImplementation(project(":AutomatticTracks"))
 }
+
+dependencyAnalysis {
+    issues {
+        onUnusedDependencies {
+            // This dependency is not needed but kept to preserve the module's default configuration.
+            exclude("org.jetbrains.kotlin:kotlin-stdlib")
+        }
+    }
+}
```
I have similar concern as in other, similar scenario. If `kotlin-stdlib` is not required, then let's maybe instruct kotlin gradle plugin to... not include `kotlin-stdlib`.

It won't affect build in any way, as this is about adding a dependency to project's classpath (not build classpath). The authors of the plugin also offered a ready solution for this: https://kotlinlang.org/docs/gradle-configure-project.html#dependency-on-the-standard-library so we don't hack anything.

ParaskP7 replied:
👋 @wzieba !

> ...then let's maybe instruct kotlin gradle plugin to... not include kotlin-stdlib.

To my understanding, you suggest adding the `kotlin.stdlib.default.dependency=false` within `gradle.properties`, but have you (maybe) tried that yourself? This, as I would also expect, fails the build, because although this `benchmark` module doesn't need the `kotlin-stdlib` dependency, all other modules do. Isn't that right? 🤔

What we want is to tell our build to ignore the `kotlin-stdlib` dependency, yes, but for `benchmark` only. I am not sure what's the correct way to do that, especially since this also needs to be done on the source-set level. 🤷 

PS: I think that for what is worth we shouldn't care too much about the configuration of this module, it is just an auto-generate benchmark module that serve a very specific purpose.

wzieba replied:
Each module can have its own `gradle.properties`. Having `kotlin.stdlib.default.dependency=false` in the `gradle.properties` in `benchmark` works just fine

ParaskP7 replied:
True @wzieba , I just tried it and it indeed works, I am curious why though, as I would have expected that `./gradlew benchmark:assembleAndroidTest` to fail since `TrackingMethodsBenchmark` is written in Kotlin, but it works and the project compiles, strange... 🤔 

No matter, this is now done: a2a349dff6166c6d9fe56b200fa80a7ff02df7b6

wzieba replied:
Thanks! Just sharing that the stdlib contains some additional features, but it's not necessary to compile Kotlin code - that's why this module still compiles.

ParaskP7 replied:
Ah true @wzieba (😅 ) , it is just a Kotlin standard library with some additional features... I didn't have to deal with that for so long that I am now forgetting things, thanks for the merge! 🤷

wzieba commented on benchmark/build.gradle.kts:
```diff
@@ -43,6 +43,11 @@ android {
         }
     }
+
+    lint {
+        warningsAsErrors = true
+        lintConfig = file("${project.rootDir}/config/lint/lint.xml")
+    }
```
I'm not sure if we should care about lint issues in a `benchmark` module, it's purely utilitarian one, used only to generate performance benchmarks when it's necessary.

ParaskP7 replied:
This is true @wzieba , but then again, during development, while I was running `./gradlew lintDebug` to verify things, Lint kept failing for me. Actually, I kept hitting that problem for months now, but never bother to do anything about it. Now, and since the configuration was very straightforward, I wanted to fix so that we Lint never fails again for this project, no matter the module.

ParaskP7 committed:
Build: Remove the kotlin stdlib default dependency from benchmark

Context: https://github.com/Automattic/
Automattic-Tracks-Android/pull/235#discussion_r1723352467

ParaskP7 committed:
Merge branch 'trunk' into build/remove-unused-dependencies
